### PR TITLE
[shard] Add ReplicatedTensor

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #73703
* __->__ #73529

Add ReplicatedTensor, a ReplicatedTensor is a type of tensor that have the same value on all ranks across the world_size.

ReplicatedTensor is a :class:`~torch.Tensor` subclass, and it could be used together with ShardedTensor/Tensor together to express different types of computation. The inter-op rules defined as (using torch.add as an example op):
    ReplicatedTensor + ReplicatedTensor = ReplicatedTensor
    ReplicatedTensor + torch.Tensor = torch.Tensor
    ReplicatedTensor + ShardedTensor = ShardedTensor

We also added a `validate()` API to help user validate if a replicated tensor on certain process_group is truly replicated or not.

TODO: next PR gonna add ShardedTensor/PartialTensor logic to handle ReplicatedTensor.

Differential Revision: [D34529374](https://our.internmc.facebook.com/intern/diff/D34529374/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34529374/)!

### CODE REVIEW

File: torch/distributed/_shard/replicated_tensor.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+import torch.distributed as dist
+
+from torch.overrides import get_default_nowrap_functions
+from torch.distributed._shard.sharded_tensor.api import ShardedTensor
+from torch.distributed import distributed_c10d
+
+
+class ReplicatedTensor(torch.Tensor):
+    """
+    ReplicatedTensor represents a tensor which is replicated across the world_size and
```
Comment (fduwjj) [inline torch/distributed/_shard/replicated_tensor.py:11]: Nit: `world_size`?

File: torch/distributed/_shard/replicated_tensor.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    has the same value on each rank.
+
+    ReplicatedTensor is a :class:`~torch.Tensor` subclass, and it could be used together
+    with ShardedTensor/Tensor together to express different types of computation. The
+    inter-op rules defined as (using torch.add as an example op):
+        ReplicatedTensor + ReplicatedTensor = ReplicatedTensor
+        ReplicatedTensor + torch.Tensor = torch.Tensor
+        ReplicatedTensor + ShardedTensor = ShardedTensor
```
Comment (fduwjj) [inline torch/distributed/_shard/replicated_tensor.py:19]: Can we also add one for _PartialTensor?

File: torch/distributed/_shard/replicated_tensor.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    def __new__(cls, data=None):
+        if data is None:
+            data = torch.empty(0)
+        return torch.Tensor._make_subclass(cls, data)
+
+    def __repr__(self):
+        return f"ReplicatedTensor({super(ReplicatedTensor, self).__repr__()})"
+
+    @classmethod
+    def __torch_function__(cls, func, types, args=(), kwargs=None):
```
Comment (fduwjj) [inline torch/distributed/_shard/replicated_tensor.py:37]: Just one n00b question here, with this override, we can enable the comparison in the test right? (Like `self.assertEqual`?)

File: torch/distributed/_shard/replicated_tensor.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+                    return v.__torch_function__(func, types, args, kwargs)
+                else:
+                    new_kwargs[k] = v
+
+        # We cann't do super().__torch_function__() as it implicitly convert the result
+        # back to tensor subclasses, where in our case, we need to control the output type
+        # base on the inter-op rules we defined.
+        with torch._C.DisableTorchFunction():
+            rs = func(*new_args, **new_kwargs)
+            if func in get_default_nowrap_functions():
```
Comment (fduwjj) [inline torch/distributed/_shard/replicated_tensor.py:78]: For learning purpose. Does this often mean the situation when it's field access like `t.grad`?

**Review by fduwjj — APPROVED** [review]
It looks great and thanks for working on this!

File: torch/distributed/_shard/replicated_tensor.py (hunk 2 above)
Comment (wanchaol) [reply torch/distributed/_shard/replicated_tensor.py:19]: I thought about it when adding the comment, I decided to leave `_PartialTensor` out because:
1. It's not a public API yet
2. Ideally PartialTensor is not sth that user need to be aware of or worry about, it's more like intermediate results handled by our internal system. So I'm worried if we add the comment here, user might be a bit confused and need to learn what a PartialTensor is?

Let me know if that make sense or not :)

File: torch/distributed/_shard/replicated_tensor.py (hunk 3 above)
Comment (wanchaol) [reply torch/distributed/_shard/replicated_tensor.py:37]: Do you mean comparison between ReplicatedTensor and Tensor, or ShardedTensor? Since `ReplicatedTensor` is a subclass of Tensor, it should work well with Tensor, have to check if it works with ShardedTensor, will try adding some tests there.

File: torch/distributed/_shard/replicated_tensor.py (hunk 4 above)
Comment (wanchaol) [reply torch/distributed/_shard/replicated_tensor.py:78]: yeah, more related to field access, this was copied from the `tensor.__torch_function__` because we don't want to go into this function as we manage our own type through the rules we defined.

File: torch/distributed/_shard/replicated_tensor.py (hunk 3 above)
Comment (wanchaol) [reply torch/distributed/_shard/replicated_tensor.py:37]: Just checked there's already assertEqual between Tensor/ReplicatedTensor, it's working, but for ReplicatedTensor/ShardedTensor, it's not yet supported, we need to add handling logic to binary_cmp, but in theory, I guess ShardedTensor will never equal to a ReplicatedTensor as they are different topology? we might need to define the rule for this here. cc @pritamdamania

File: torch/distributed/_shard/api.py [hunk 5]
```diff
@@ -118,3 +119,15 @@ def shard_parameter(
 
     # Now we can set the attribute appropriately.
     setattr(module, param_name, st)
+
+
+def _replicate_tensor(tensor: torch.Tensor) -> ReplicatedTensor:
+    """
+    Given a :class:`torch.Tensor`, mark it as a ReplicatedTensore where all
```
Comment (pritamdamania87) [inline torch/distributed/_shard/api.py:126]: nit: s/ReplicatedTensore/ReplicatedTensor

File: torch/distributed/_shard/api.py [hunk 6]
```diff
@@ -118,3 +119,15 @@ def shard_parameter(
 
     # Now we can set the attribute appropriately.
     setattr(module, param_name, st)
+
+
+def _replicate_tensor(tensor: torch.Tensor) -> ReplicatedTensor:
```
Comment (pritamdamania87) [inline torch/distributed/_shard/api.py:124]: Do we really need an API like this? Users can just call `ReplicatedTensor(tensor)`?

File: torch/distributed/_shard/replicated_tensor.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    inter-op rules defined as (using torch.add as an example op):
+        ReplicatedTensor + ReplicatedTensor = ReplicatedTensor
+        ReplicatedTensor + torch.Tensor = torch.Tensor
+        ReplicatedTensor + ShardedTensor = ShardedTensor
+
+    NOTE: We do not gurantee equal content of ReplicatedTensor across nodes after its
+    construction. Although we defined proper inter-op rules to make sure ReplicatedTensor
+    stays the same, there's no enforcement on it (i.e. if you manually modify content on
+    some ranks, the modified value will not automatically get synced to other nodes). If
+    you wish to manually validate tensors are the same across ranks, use `validate()`.
```
Comment (pritamdamania87) [inline torch/distributed/_shard/replicated_tensor.py:25]: Not for this PR, but I think we need two modes for ReplicatedTensor. The first mode is what we have here where it is just a tag to help sharded computations, but probably this should not be the default mode.

I think the default mode should be similar to DDP, where ReplicatedTensor broadcasts the torch.Tensor on rank 0 (probably can also be optionally specified which rank). Then in the backward pass for this mode, we always allreduce the gradients for the ReplicatedTensor. This means ReplicatedTensor can stand on its own.

File: torch/distributed/_shard/replicated_tensor.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+    """
+    def __new__(cls, data=None):
```
Comment (pritamdamania87) [inline torch/distributed/_shard/replicated_tensor.py:28]: ReplicatedTensor should take an optional process_group indicating the replication environment.

File: torch/distributed/_shard/replicated_tensor.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+                    # TODO: handle ShardedTensor inter-op with ReplicatedTensor
+                    return v.__torch_function__(func, types, args, kwargs)
+                else:
+                    new_kwargs[k] = v
+
+        # We cann't do super().__torch_function__() as it implicitly convert the result
+        # back to tensor subclasses, where in our case, we need to control the output type
+        # base on the inter-op rules we defined.
+        with torch._C.DisableTorchFunction():
+            rs = func(*new_args, **new_kwargs)
```
Comment (pritamdamania87) [inline torch/distributed/_shard/replicated_tensor.py:77]: Why do we need `new_args` and `new_kwargs` here? Can't we jus pass in args and kwargs?

File: torch/distributed/_shard/sharded_tensor/__init__.py [hunk 10]
```diff
@@ -366,7 +366,7 @@ def sharded_op_impl(func):
     parameters, the function provided will be invoked for that operator.
 
     Example::
-        >>> @custom_sharded_op(torch.nn.functional.linear)
+        >>> @sharded_op_impl(torch.nn.functional.linear)
```
Comment (pritamdamania87) [inline torch/distributed/_shard/sharded_tensor/__init__.py:369]: Thanks for fixing!

File: torch/distributed/_shard/replicated_tensor.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+
+        if kwargs is not None:
+            for k, v in kwargs.items():
+                if isinstance(v, ShardedTensor):
+                    # redispatch to ShardedTensor
+                    # TODO: handle ShardedTensor inter-op with ReplicatedTensor
+                    return v.__torch_function__(func, types, args, kwargs)
+                else:
+                    new_kwargs[k] = v
```
Comment (pritamdamania87) [inline torch/distributed/_shard/replicated_tensor.py:71]: Why can't we put this logic into the loops for args and kwargs above?

File: torch/distributed/_shard/replicated_tensor.py [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+        # back to tensor subclasses, where in our case, we need to control the output type
+        # base on the inter-op rules we defined.
+        with torch._C.DisableTorchFunction():
+            rs = func(*new_args, **new_kwargs)
+            if func in get_default_nowrap_functions():
+                return rs
+            if not has_tensor and isinstance(rs, torch.Tensor) and not isinstance(rs, cls):
+                # if it does not have tensor operands and does not get dispatched to ShardedTensor
+                # __torch_function__, we return a ReplicatedTensor according to our inter-op rule
+                rs = rs.as_subclass(cls)
```
Comment (pritamdamania87) [inline torch/distributed/_shard/replicated_tensor.py:83]: Does this cover all cases? What happens if we add a scalar to ReplicatedTensor (which is different on each rank)? Doesn't this check pass and we incorrectly create a ReplicatedTensor?

File: torch/distributed/_shard/replicated_tensor.py [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+
+            return rs
+
+    def validate(self, process_group=None) -> bool:
```
Comment (pritamdamania87) [inline torch/distributed/_shard/replicated_tensor.py:87]: process_group should be passed in during construction of ReplicatedTensor.

File: torch/distributed/_shard/replicated_tensor.py [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+                    new_kwargs[k] = v
+
+        # We cann't do super().__torch_function__() as it implicitly convert the result
+        # back to tensor subclasses, where in our case, we need to control the output type
+        # base on the inter-op rules we defined.
+        with torch._C.DisableTorchFunction():
+            rs = func(*new_args, **new_kwargs)
+            if func in get_default_nowrap_functions():
+                return rs
+            if not has_tensor and isinstance(rs, torch.Tensor) and not isinstance(rs, cls):
```
Comment (pritamdamania87) [inline torch/distributed/_shard/replicated_tensor.py:80]: If we are adding two ReplicatedTensors here, shouldn't we validate they are on the same PG? I feel this check might be more clear if we assert all args are ReplicatedTensor and only in that case we return a ReplicatedTensor using `rs.as_subclass`. In all other cases, we return rs.

File: test/distributed/_shard/test_replicated_tensor.py [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+    @with_comms(init_rpc=False)
+    @skip_if_lt_x_gpu(4)
+    @requires_nccl()
+    def test_replicated_tensor_basics(self):
+        local_tensor = torch.ones(3, 3, device=f"cuda:{self.rank}") * 4
+        replica_tensor = ReplicatedTensor(local_tensor)
+        # validate it's a replicated tensor by checking values on all rank
+        validated = replica_tensor.validate()
+        self.assertEqual(validated, True)
+        self.assertEqual(replica_tensor + 2, torch.ones(3, 3) * 6)
```
Comment (pritamdamania87) [inline test/distributed/_shard/test_replicated_tensor.py:26]: nit: validate type of `replica_tensor + 2`

File: test/distributed/_shard/test_replicated_tensor.py [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+    @requires_nccl()
+    def test_replicated_tensor_inter_op_replicated_tensor(self):
+        local_tensor = torch.ones(3, 3, device=f"cuda:{self.rank}")
+        replica_tensor1 = ReplicatedTensor(local_tensor * 4)
+        replica_tensor2 = ReplicatedTensor(local_tensor * 6)
+
+        new_tensor = replica_tensor1 * replica_tensor2
+        self.assertTrue(isinstance(new_tensor, ReplicatedTensor))
+        self.assertEqual(new_tensor, torch.ones(3, 3) * 24)
+
```
Comment (pritamdamania87) [inline test/distributed/_shard/test_replicated_tensor.py:46]: This should work only if PGs are same.

File: test/distributed/_shard/test_replicated_tensor.py [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+    @with_comms(init_rpc=False)
+    @skip_if_lt_x_gpu(4)
+    @requires_nccl()
+    def test_replicated_tensor_inter_op_tensor(self):
+        local_tensor = torch.ones(3, 3, device=f"cuda:{self.rank}") * 4
+        replica_tensor = ReplicatedTensor(local_tensor)
+
```
Comment (pritamdamania87) [inline test/distributed/_shard/test_replicated_tensor.py:53]: Are these two interops coming in follow up PRs?

1. ShardedTensor + ReplicatedTensor
2. PartialTensor + ReplicatedTensor

File: torch/distributed/_shard/replicated_tensor.py (hunk 12 above)
Comment (wanchaol) [reply torch/distributed/_shard/replicated_tensor.py:83]: Yeah agreed, should we just enforce the rule that Anything else inter-op with ReplicatedTensor, will result in the type not being ReplicatedTensor anymore? That might make our life easier as we cannot ensure the operand to be the same unless it marked as replicated.

File: test/distributed/_shard/test_replicated_tensor.py (hunk 15 above)
Comment (wanchaol) [reply test/distributed/_shard/test_replicated_tensor.py:26]: I guess the type of `replica_tensor + 2` should becomes `tensor` instead of replicated tensor

File: torch/distributed/_shard/api.py (hunk 6 above)
Comment (wanchaol) [reply torch/distributed/_shard/api.py:124]: My original intention on introducing this API is to make sure we have consistent APIs provided to the user. We now have:
1. shard_module(module,plan)
2. shard_parameter(module, param_name, spec)
3. _shard_tensor(tensor, spec)
All these three APIs are being used to mark a param or tensor as ShardedTensor, I feel that we should have a similar API to mark a tensor as ReplicatedTensor, it makes the API more consistent from user prospective. Let me know if this make sense or we could just use `ReplicatedTensor(tensor)`

File: torch/distributed/_shard/replicated_tensor.py (hunk 8 above)
Comment (wanchaol) [reply torch/distributed/_shard/replicated_tensor.py:28]: Yeah I agree we should have ReplicatedTensor tight with a replication env (a pg), but I didn't do this initially bc it requires the Tensor to hold a metadata as its member, which `torch.Tensor.make_subclass` does not work in that way. Let me see if I can change it to `make_wrapper_subclass` and define `__torch_dispatch__` together with `__torch_function__`

File: torch/distributed/_shard/replicated_tensor.py (hunk 2 above)
Comment (fduwjj) [reply torch/distributed/_shard/replicated_tensor.py:19]: Got it, makes sense.

File: torch/distributed/_shard/replicated_tensor.py (hunk 8 above)
Comment (wanchaol) [reply torch/distributed/_shard/replicated_tensor.py]: I actually found a way to propagate the field even with `make_subclass`, so no need to define `__torch_dispatch__` yet, just updated the PR

File: torch/distributed/_shard/replicated_tensor.py (hunk 12 above)
Comment (pritamdamania87) [reply torch/distributed/_shard/replicated_tensor.py]: > Anything else inter-op with ReplicatedTensor, will result in the type not being ReplicatedTensor anymore?

Yes, I think that is a safer rule.

File: torch/distributed/_shard/replicated_tensor.py [hunk 18]
```diff
@@ -25,66 +26,85 @@ class ReplicatedTensor(torch.Tensor):
...
+                    if replicated_pg is None:
+                        replicated_pg = v.process_group
+                    elif replicated_pg != v.process_group:
+                        raise RuntimeError(
+                            f"ReplicatedTensor operands must be in the same process group "
+                            f"in torch function '{func.__name__}', but found at least two "
+                            f"ReplicatedTensor operands in different process groups! ")
                 else:
-                    new_kwargs[k] = v
+                    all_replicated = False
```
Comment (pritamdamania87) [inline torch/distributed/_shard/replicated_tensor.py:88]: nit: repeated code between args and kwargs, maybe create a simple inline helper function and dedup this.

**Review by pritamdamania87 — APPROVED** [review]
Looks good! Just a minor comment regarding some refactoring.
